### PR TITLE
Adding Private Range (No SNAT) Support to Firewall Policies

### PR DESCRIPTION
**Pull request by kenchenMicrosoft**
## Description

Add NoSnat Private Range for Firewall Policies

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/816
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs [hunk 1]
```diff
@@ -132,6 +132,12 @@ public class NewAzureFirewallPolicyCommand : AzureFirewallPolicyBaseCmdlet
         [ValidateNotNullOrEmpty]
         public PSManagedServiceIdentity Identity { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "The private IP ranges to which traffic won't be SNAT'ed"
+        )]
+        public string[] PrivateRange { get; set; }
```
Comment (BethanyZhou) [inline src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs:139]: I think the name of PrivateIpRange is more specific for this parameter, what's your thought?

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs [hunk 2]
```diff
@@ -42,5 +45,77 @@ public class PSAzureFirewallPolicy : PSTopLevelResource
...
+            if (split.Length != 2)
+                throw new PSArgumentException(String.Format("\'{0}\' is not a valid private range ip address", ipAddress));
+
+            // validate the ip
+            ValidateSingleIpAddress(split[0]);
+
+            // validate mask
+            var bit = 0;
+            if (!Int32.TryParse(split[1], out bit) || bit < 0 || bit > 32)
+                throw new PSArgumentException(String.Format("\'{0}\' is not a valid private range ip address, subnet mask should between 0 and 32", ipAddress));
```
Comment (BethanyZhou) [inline src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs:109]: Suggested change P
SArgumentException into AzPSArgumentException

File: src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs (hunk 1 above)
Comment (kenchenMicrosoft) [reply src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs:139]: I think that makes sense. However, AzureFirewall already has this PrivateRange field defined for some time now. That's why AzureFirewallPolicy is naming it the same way to be consistent. Here is the design review for it:
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/816

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs (hunk 2 above)
Comment (kenchenMicrosoft) [reply src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs]: Updated accordingly

**Comment by kenchenMicrosoft** [comment]
@BethanyZhou the build failed in Linux. The error doesn't seem to be related to my code changes. Do you know what could be the cause?

https://dev.azure.com/azure-sdk/public/_build/results?buildId=790931&view=logs&j=2f953adc-c56d-55c4-a64a-eab7c4b02235&t=fc7ea605-a507-5208-bc88-3e6a658c906b&l=100

File: src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs (hunk 1 above)
Comment (BethanyZhou) [reply src/Network/Network/AzureFirewallPolicy/NewAzureFirewallPolicyCommand.cs:139]: Fine, got it.

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs [hunk 3]
```diff
@@ -42,5 +47,77 @@ public class PSAzureFirewallPolicy : PSTopLevelResource
...
+                    ValidateSingleIpAddress(ip);
+            }
+        }
+
+        private void ValidateSingleIpAddress(string ipAddress)
+        {
+            IPAddress ipVal;
+            if (!IPAddress.TryParse(ipAddress, out ipVal) || ipVal.AddressFamily != System.Net.Sockets.AddressFamily.InterNetwork)
+            {
+                throw new AzPSArgumentException(String.Format("\'{0}\' is not a valid private range ip address", ipAddress), "IpAddress", ErrorKind.UserError);
```
Comment (BethanyZhou) [inline src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs:95]: It's encouraged that put these exception messages into azure-powershell/src/Network/Network/Properties/Resources.resx and refer them by `Resources.{ErrorName}`
sth like ```throw new AzPSArgumentException(String.Format("Resources.InvalidIpAddress, ipAddress), nameof(IpAddress), ErrorKind.UserError);```
```

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs [hunk 4]
```diff
@@ -42,5 +47,77 @@ public class PSAzureFirewallPolicy : PSTopLevelResource
...
+            var bit = 0;
+            if (!Int32.TryParse(split[1], out bit) || bit < 0 || bit > 32)
+                throw new AzPSArgumentException(string.Format("\'{0}\' is not a valid private range ip address, subnet mask should between 0 and 32", ipAddress), "IpAddress", ErrorKind.UserError);
+
+            // validated that unmasked bits are 0
+            var splittedIp = split[0].Split('.');
+            var ip = Int32.Parse(splittedIp[0]) << 24;
+            ip = ip + Int32.Parse(splittedIp[1]) << 16 + Int32.Parse(splittedIp[2]) << 8 + Int32.Parse(splittedIp[3]);
+            if (ip << bit != 0)
+                throw new AzPSArgumentException(String.Format("\'{0}\' is not a valid private range ip address, bits not covered by subnet mask should be all 0", ipAddress), "IpAddress", ErrorKind.UserError);
```
Comment (BethanyZhou) [inline src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs:118]: It's encouraged that put these exception messages into azure-powershell/src/Network/Network/Properties/Resources.resx

**Comment by BethanyZhou** [comment]
> @BethanyZhou Beisi Zhou FTE the build failed in Linux. The error doesn't seem to be related to my code changes. Do you know what could be the cause?
> 
> https://dev.azure.com/azure-sdk/public/_build/results?buildId=790931&view=logs&j=2f953adc-c56d-55c4-a64a-eab7c4b02235&t=fc7ea605-a507-5208-bc88-3e6a658c906b&l=100

Let's see what happen if re-run this pipeline

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs (hunk 3 above)
Comment (kenchenMicrosoft) [reply src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs]: Updated accordingly

File: src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs (hunk 4 above)
Comment (kenchenMicrosoft) [reply src/Network/Network/Models/AzureFirewallPolicy/PSAzureFirewallPolicy.cs]: Updated accordingly

**Review by BethanyZhou — DISMISSED** [review]
LGTM

**Comment by BethanyZhou** [comment]
Hi @kenchenMicrosoft , all features in branch network-november is ready to be released tomorrow and branch network-november is about to be deprecated. Please target this pr to branch network-february tomorrow.

**Comment by kenchenMicrosoft** [comment]
Thanks @BethanyZhou!  Is there anyway I can squeeze this PR in the next hotfix release? Do you know who I can talk to make a request for a hotfix release?

**Comment by BethanyZhou** [comment]
Hi @kenchenMicrosoft , what's your ddl for this feature? Our next official release is [S185](https://github.com/Azure/azure-powershell/milestone/120) on 04/13. Is it good for you? Otherwise, please ask @dingmeng-xue to approve a OOB release.

**Comment by dingmeng-xue** [comment]
Please target the PR to branch network-february

File: src/Network/Network/ChangeLog.md [hunk 5]
```diff
@@ -38,6 +38,10 @@
 	- `Get-AzVpnClientConfiguration`
 	- `New-AzVirtualNetworkGateway`
 	- `Set-AzVirtualNetworkGateway`
+* Updated cmdlets to enable setting of PrivateRange on AzureFirewallPolicy.
+    - `New-AzFirewallPolicy`
+    - `Set-AzFirewallPolicy`
+
```
Comment (BethanyZhou) [inline src/Network/Network/ChangeLog.md:44]: Hi @kenchenMicrosoft , the log msgs should be added under the top of ## Upcoming Release section. Please merge network-feb into your branch, then you will detect that there should be a conflict.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Network/Network/ChangeLog.md (hunk 5 above)
Comment (kenchenMicrosoft) [reply src/Network/Network/ChangeLog.md]: updated accordingly

**Review by BethanyZhou — APPROVED** [review]

**Pull request merged**